verilog/formatting/formatter.cc

```diff
@@ -230,15 +230,265 @@ static void DeterminePartitionExpansion(partition_node_type* node,
+// This function creates collection of unwrapped lines from vector tree and
+// calls AppendFittingSubpartitions to format it according to
+// AppendFittingSubpartitions algorithm:
+// 1. Decides whether or not to wrap the second line, a binary decision
+// 2. To weigh this decision, evaluate both branches in a simple greedy manner,
+//    that is by appending consecutive lines and wrapping only when necessary to stay
+//    under the column limit (function AppendFittingSubpartitions()).
+// 3. Chooses branch with fewer lines, or unwrapped branch if number of lines is equal
+//
+// For example:
+//
+// 1. Prefers first one (less lines)
+// pretend column limit here           ->|
+// [function fffffffffff(]
+// [>>>>type_a aaaa, type_b bbbbb,]
+// [>>>>type_c cccccc, type_d dddddddd,]
+// [>>>>type_e eeeeeeee, type_f ffff);]
+//
+// pretend column limit here           ->|
+// [function fffffffffff(type_a aaaa,]
+// [>>>>>>>>>>>>>>>>>>>>type_b bbbbb,]
+// [>>>>>>>>>>>>>>>>>>>>type_c cccccc,]
+// [>>>>>>>>>>>>>>>>>>>>type_d dddddddd,]
+// [>>>>>>>>>>>>>>>>>>>>type_e eeeeeeee,]
+// [>>>>>>>>>>>>>>>>>>>>type_f ffff);]
+//
+// 2. Prefers first one (same number of lines, prefers unwrapped result)
+// pretend column limit here                           ->|
+// [function fffffffffff(type_a aaaa, type_b bbbbb,]
+// [>>>>>>>>>>>>>>>>>>>>>type_c cccccc, type_d dddddddd,]
+// [>>>>>>>>>>>>>>>>>>>>>type_e eeeeeeee, type_f ffff);]
+//
+// pretend column limit here                           ->|
+// [function fffffffffff(]
+// [>>>>type_a aaaa, type_b bbbbb, type_c cccccc,]
+// [>>>>type_d dddddddd, type_e eeeeeeee, type_f ffff);]
+//
+static void ReshapeFittingSubpartitions(partition_node_type& node,
+                                        const FormatStyle& style) {
+  auto& node_view = node.Value();
+  auto& children = node.Children();
+  const auto& uwline = node_view.Value();
+  const auto partition_policy = uwline.PartitionPolicy();
+
+  // Handle only kAppendFittingSubpartitions policy, ignore rest of them
+  switch (partition_policy) {
+  case PartitionPolicyEnum::kAppendFittingSubPartitions:
+    break;
+
+  default:
+    return;
+  }
+
+  // Check if this partition fits in line
+  // If so, Unexpand() and don't waste time trying to fit it
+  // If not, force Expand()
+  if (verible::FitsOnLine(uwline, style).fits) {
+    node_view.Unexpand();
+    return ;
+  } else {
+    // if not, force expansion and apply alogrithm
+    node_view.Expand();
+  }
+
+  //
+  // One-level partition tree expansion
+  // (ignores subpartition policies)
+  //
+  // e.g. input:
+  // -----------
+  // function fffffffffff(type_a aaaa, type_b bbbbb);
+  //
+  // partition tree:
+  // ---------------
+  // { ([<auto>]) @{2}, policy: append-fitting-sub-partitions
+  //   { ([function fffffffffff (]) }
+  //   { (>>>>[<auto>]) @{2,1}, policy: fit-else-expand
+  //     { (>>>>[type_a aaaa ,]) }
+  //     { (>>>>[type_b bbbbb ,]) }
```

Comment:
I think this should be: `{ (>>>>[type_b bbbbb ; )]) }` (and include the following `);`.

Reply:
Merged.